engineapi/proxy.go

```diff
@@ -1,288 +1,128 @@
-	if im == nil {
-		return nil, errors.Errorf("BUG: failed to get proxy client due to missing instance manager")
-	}
-
-	log := getLoggerForProxy(h.logger, im.Spec.Image)
-
-	h.lock.Lock()
-	client, exist := h.Clients[im.Spec.Image]
-	h.lock.Unlock()
-	if exist {
-		err := client.Ping()
-		if err == nil {
-			return client, nil
-		}
-
-		log.WithError(err).Debug("Resetting proxy connection")
-		client.Stop(im.Spec.Image)
-
-		h.lock.Lock()
-		delete(h.Clients, im.Spec.Image)
-		h.lock.Unlock()
+		return nil, errors.Errorf("BUG: invalid engine client proxy fallback client: %v", fallBack)
 	}
 
-	return h.NewClient(log, im)
+	return NewEngineClientProxy(im, ds, log)
 }
 
-func (h *EngineClientProxyHandler) NewClient(log logrus.FieldLogger, im *longhorn.InstanceManager) (EngineClientProxy, error) {
-	h.lock.Lock()
-	defer h.lock.Unlock()
-
-	if existClient, exist := h.Clients[im.Spec.Image]; exist {
-		// TODO: we do not delete the map to avoid nil memory panic, so need to
-		// check the ping to find any concurrent NewClient. Fix this when
-		// refactoring to use goroutine for proxy client handling.
-		err := existClient.Ping()
-		if err == nil {
-			log.Debugf("Proxy client already created")
-			return existClient, nil
-		}
-	}
+func NewEngineClientProxy(im *longhorn.InstanceManager, ds *datastore.DataStore, logger logrus.FieldLogger) (c EngineClientProxy, err error) {
+	defer func() {
+		err = errors.Wrap(err, "failed to get engine client proxy")
+	}()
 
 	isInstanceManagerRunning := im.Status.CurrentState == longhorn.InstanceManagerStateRunning
 	if !isInstanceManagerRunning {
-		return nil, errors.Errorf("cannot create new gRPC client while instance manager in %v state", im.Status.CurrentState)
+		err = errors.Errorf("%v instance manager is in %v, not running state", im.Name, im.Status.CurrentState)
+		return nil, err
 	}
 
 	hasIP := im.Status.IP != ""
 	if !hasIP {
-		return nil, errors.Errorf("cannot create new gRPC client while instance manager status IP is empty")
+		err = errors.Errorf("%v instance manager status IP is missing", im.Name)
+		return nil, err
 	}
 
-	client := &Proxy{}
-	if err := client.Start(im, log, h.ds); err != nil {
+	ctx, cancel := context.WithCancel(context.Background())
+	client, err := imclient.NewProxyClient(ctx, cancel, im.Status.IP, InstanceManagerProxyDefaultPort)
+	if err != nil {
 		return nil, err
 	}
 
-	log.Debug("Adding client to proxy handler")
-	h.Clients[im.Spec.Image] = client
-
-	return client, nil
+	return &Proxy{
+		logger:     logger,
+		grpcClient: client,
+		ds:         ds,
+	}, nil
 }
 
 type Proxy struct {
-	logger logrus.FieldLogger
-
-	ds *datastore.DataStore
-
-	InstanceManagerName         string
-	InstanceManagerControllerID string
-
-	// used to notify the controller that proxy has stopped
-	voluntaryStopCh chan struct{}
-	stopCh          chan struct{}
-
+	logger     logrus.FieldLogger
+	ds         *datastore.DataStore
```

Comment:
It seems that the proxy does not need this

Reply:
Removed.